tests/test_dynamic_publisher_block.py

```diff
@@ -82,46 +82,46 @@ def test_reusing_pubs(self):
             self.assertEqual(pub.return_value.open.call_count, 1)
             pub.return_value.send.assert_called_with(signals)
 
-    def test_partitioning(self):
-        publisher = DynamicPublisher()
+    @patch(DynamicPublisher.__module__ + '.Publisher', side_effect=[Mock(), Mock(), Mock()])
+    def test_partitioning(self, pub):
+        block = DynamicPublisher()
         topic = "topic.{{ $sig }}"
 
-        foo_pub = Mock()
-        bar_pub = Mock()
-        baz_pub = Mock()
-
-        with patch(DynamicPublisher.__module__ + '.Publisher', side_effect=[foo_pub, bar_pub, baz_pub]) as pub:
-            self.configure_block(publisher, {"topic": topic})
-            publisher.start()
-
-            signals = [
-                Signal(dict(sig="foo", val=1)),
-                Signal(dict(sig="bar", val=2)),
-                Signal(dict(sig="baz", val=3)),
-                Signal(dict(sig="foo", val=4)),
-                Signal(dict(sig="bar", val=5)),
-                Signal(dict(sig="foo", val=6)),
-            ]
-            publisher.process_signals(signals)
-
-            pub.assert_any_call(topic="topic.foo")
-            pub.assert_any_call(topic="topic.bar")
-            pub.assert_any_call(topic="topic.baz")
-
-            foo_pub.send.assert_called_once_with([
-                Signal(dict(sig="foo", val=1)),
-                Signal(dict(sig="foo", val=4)),
-                Signal(dict(sig="foo", val=6)),
-            ])
-
-            bar_pub.send.assert_called_once_with([
-                Signal(dict(sig="bar", val=2)),
-                Signal(dict(sig="bar", val=5)),
-            ])
-
-            baz_pub.send.assert_called_once_with([
-                Signal(dict(sig="baz", val=3)),
-            ])
+        self.configure_block(block, {"topic": topic})
+        block.start()
+
+        signals = [
+            Signal(dict(sig="foo", val=1)),
+            Signal(dict(sig="bar", val=2)),
+            Signal(dict(sig="baz", val=3)),
+            Signal(dict(sig="foo", val=4)),
+            Signal(dict(sig="bar", val=5)),
+            Signal(dict(sig="foo", val=6)),
+        ]
+        block.process_signals(signals)
+
+        pub.assert_any_call(topic="topic.foo")
+        pub.assert_any_call(topic="topic.bar")
+        pub.assert_any_call(topic="topic.baz")
+
+        foo_pub, _ = block._cache.get("topic.foo", Mock())
```

Comment:
I would probably remove the default value of `Mock()` from these cache retrieval lines. It could potentially suppress a problem in the block or a typo somewhere. You should expect all 3 topics to hit the cache I think.

Reply:
I put these here _because_ I am reaching into something that i probably shouldn't for the test, and didn't want the defaultdict to _actually_ invoke the `default_factory` method by accident, incurring its side-effect. _But_, now that I look at the documentation for `.get()`, and how its different than `[]`, then it shouldn't be invoked in that case so its safe.